From e33c46e91b584cada45b16cb1856680e301bbd3d Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:44 -0400
Subject: [fs] nfsd: nfsd4_process_open2() must reference the open stateid

Message-id: <1410898849-22126-101-git-send-email-bfields@redhat.com>
Patchwork-id: 95788
O-Subject: [Patch RHEL7 100/165] nfsd: nfsd4_process_open2() must reference the open stateid
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Ensure that nfsd4_process_open2() keeps a reference to the open
stateid until it is done working with it. Necessary step toward
client_mutex removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: d6f2bc5dcf58259b6c3f206ae8f14087300b5bcf
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 5cb6305..f3018cb 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -2996,6 +2996,7 @@ alloc_init_open_stateowner(unsigned int strhashval, struct nfsd4_open *open,
 static void init_open_stateid(struct nfs4_ol_stateid *stp, struct nfs4_file *fp, struct nfsd4_open *open) {
  struct nfs4_openowner *oo = open->op_openowner;
 
+ atomic_inc(&stp->st_stid.sc_count);
  stp->st_stid.sc_type = NFS4_OPEN_STID;
  INIT_LIST_HEAD(&stp->st_locks);
  stp->st_stateowner = &oo->oo_owner;
@@ -3376,6 +3377,7 @@ nfsd4_find_existing_open(struct nfs4_file *fp, struct nfsd4_open *open)
    continue;
   if (local->st_stateowner == &oo->oo_owner) {
    ret = local;
+   atomic_inc(&ret->st_stid.sc_count);
    break;
   }
  }
@@ -3836,6 +3838,8 @@ out:
   open->op_rflags |= NFS4_OPEN_RESULT_CONFIRM;
  if (dp)
   nfs4_put_stid(&dp->dl_stid);
+ if (stp)
+  nfs4_put_stid(&stp->st_stid);
 
  return status;
 }
-- 
1.7.1